source/common/http/conn_manager_impl.cc

```diff
@@ -1126,6 +1192,40 @@ void ConnectionManagerImpl::ActiveStreamEncoderFilter::
 
 void ConnectionManagerImpl::ActiveStreamEncoderFilter::continueEncoding() { commonContinue(); }
 
+void ConnectionManagerImpl::ActiveStreamEncoderFilter::responseDataTooLarge() {
+  if (parent_.encoder_filters_streaming_) {
+    onEncoderFilterAboveWriteBufferHighWatermark();
+  } else {
+    // If headers have not been sent to the user, send a 500.
+    if (!headers_continued_) {
+      // Make sure we won't end up with nested watermark calls from the body buffer.
+      parent_.encoder_filters_streaming_ = true;
+      stopped_ = false;
+      HeaderMapPtr response_headers{new HeaderMapImpl{
+          {Headers::get().Status, std::to_string(enumToInt(Http::Code::InternalServerError))}}};
+      std::string body_text = CodeUtility::toString(Http::Code::InternalServerError);
+      response_headers->insertContentLength().value(body_text.size());
+      response_headers->insertContentType().value(Headers::get().ContentTypeValues.Text);
+
+      parent_.response_headers_ = std::move(response_headers);
+      // Bypass the filters and send the response headers and body directly to the encoder.
```

Comment:
This section is scarily fragile since it's uncommon code paths.  I really wanted strong integration tests but it turns out (after wasting a bunch of time getting an end to end json transcoder filter test running) that the only filter which delays headers and has the connection manager buffer body is the http bridge filter (already tested, caught at one bug)

n.b.  if anyone adds a filter which delays sending all data via connection manager we should make the author add HTTP/H2 integration tests for it backing up.

Reply:
I assume you have conn_manager_impl tests for this case though?

Reply:
Yep!  unit and integration.  I'm more worried about integration though - it was bits of the system not playing with other bits I'm more concerned about and the tests mock out enough state they missed one bug which the integration test caught.